### PR TITLE
Implement the service endpoints controller

### PR DESCRIPTION
**Pull request by mkeeler**
### Description

This PR builds on #17215. The workload mapper tracks which services each workload is a part of. When reconciling service endpoints, the controller will aggregate view all the status information of workloads in order to populate the HealthStatus field on the ServiceEndpoints. The final result is that the ServiceEndpoints types get computed for all services.

To make the workload to service mapping more efficient it uses a radix tree. I copied this from github.com/armon/go-radix and then modified that to be a generic type. Mainly I didn't want to type coerce from interface{} all over the place. This could probably also have used a generic version of github.com/hashicorp/go-immutable-radix.S

- [ ] Unit tests

### Testing & Reproduction steps

* So far this has just involved manual testing by writing resources and observing the behavior.

### PR Checklist

* [ ] updated test coverage
* [ ] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern

### CODE REVIEW

File: internal/catalog/internal/controllers/endpoints/controller.go
Comment (boxofrad) [inline internal/catalog/internal/controllers/endpoints/controller.go:58]: As an aside, I think it would be acceptable under the Go guidance to have a context on this struct as long as you're only using it as a method parameter and not storing it long-term.

Another pattern I've seen before (see [this Slack thread](https://go.hashi.co/go-context-embedding) for pros/cons) is to _embed_ a context in your struct so you can pass it around, to the gRPC client for example, as if it were a regular context.

```Go
type myContext struct {
    context.Context

    someOtherField string
}

func doSomething(ctx myContext, rt controller.Runtime) {
    rsp, err := rt.Client.Read(ctx, ...)
}
```

Folks in the thread were pretty polarised on this though. So maybe best avoided!

File: internal/radix/doc.go
Comment (boxofrad) [inline internal/radix/doc.go:4]: Is the plan to contribute this upstream? Perhaps as a `v2` of the module? Or keep it in-tree, if you'll excuse the pun 🌴 😄

File: internal/catalog/internal/controllers/endpoints/workload_mapper.go
Comment (analogue) [inline internal/catalog/internal/controllers/endpoints/workload_mapper.go:23]: L22 and L50: update to "Services associated with a Workload"

File: internal/radix/doc.go
Comment (mkeeler) [reply internal/radix/doc.go:4]: I actually have a branch with all these changes in it. I kind of didn't want to open that can of worms yet. I could see publishing the changes to github.com/hashicorp/go-radix (out of the armon namespace).

File: internal/catalog/internal/controllers/endpoints/controller.go
Comment (ishustava) [inline internal/catalog/internal/controllers/endpoints/controller.go:37]: One potential other edge case I've been thinking through with this is when the server restarts. In that case, when the server restarts, we will:
1. Start the controller. At this point, we already have some data in the store but the workload mapper is not yet populated because we haven't reconciled anything. 
2. Running the controller will start the watches for all types asynchronously.
3. It may be possible that by the time we need to map the workload we still haven't reconciled any services or service endpoints pointing to it, and so they will not be in the mapper. It would probably be affected by the number of things we need to add to the queue as well.

I think this is similar to the case we chatted about offline earlier today, but I don't think the order of calls in the reconcile will fix it completely because maybe the `Reconcile` that would populate the mapper hasn't been called yet.

Maybe for that case, we need some sort of fall-back mechanism so that we don't miss events in case the mapper is empty (like when there's a cache miss). So if the mapper doesn't return any requests, then maybe we'd do the naive check and look through all services.

**Review by ishustava — COMMENTED** [review]
I'm still reviewing but thought I'd leave comments I had so far!

File: internal/catalog/internal/mappers/selectiontracker/selection_tracker.go
Comment (mkeeler) [reply internal/catalog/internal/mappers/selectiontracker/selection_tracker.go:58]: The controllers queues do this for us anyways so I don't see much reason in preemptively doing it.

File: internal/catalog/internal/controllers/endpoints/controller.go
Comment (mkeeler) [reply internal/catalog/internal/controllers/endpoints/controller.go]: I ended up going a different way that didn't use this object so the whole issue was avoided.

File: internal/radix/doc.go
Comment (mkeeler) [reply internal/radix/doc.go:4]: I have PR'ed it upstream. We will see if Armon wants to merge it: https://github.com/armon/go-radix/pull/23

File: internal/catalog/internal/controllers/endpoints/controller.go
Comment (ishustava) [inline internal/catalog/internal/controllers/endpoints/controller.go:63]: Should we add some logging to this controller as well?

**Review by ishustava — APPROVED** [review]
Looks great! I had a few more comments, but won't block the merge on those.

File: internal/catalog/internal/controllers/endpoints/controller.go
Comment (mkeeler) [reply internal/catalog/internal/controllers/endpoints/controller.go:63]: Yep, I can do that.

File: internal/catalog/internal/controllers/endpoints/controller.go
Comment (mkeeler) [reply internal/catalog/internal/controllers/endpoints/controller.go:288]: Where do you think we should report that. Presumably on the service or service endpoints?

Comment (ishustava) [reply internal/catalog/internal/controllers/endpoints/controller.go:288]: I think on the service because that's where user provided invalid config.

Comment (mkeeler) [reply internal/catalog/internal/controllers/endpoints/controller.go:288]: Is it the service thats invalid though?

If 9/10 workloads have that port with the correct protocol but 1/10 has the mismatch we could argue that its the workload thats the problem.

I think adding the status to the workload isn't right as the workload shouldn't need to care about service concerns in theory.

So adding it to the service could be fine but I could also see this being a status on the endpoints.

Comment (ishustava) [reply internal/catalog/internal/controllers/endpoints/controller.go:288]: That's fair. I still think in the case you mentioned it could also be the service that's misconfigured because maybe you've selected a workload that shouldn't be part of the service. Ultimately, we don't know what the user's intention was and so saying that something is misconfigured on the service makes sense to me even if it's the workload port that needs to change in the end. 

We could maybe add it to both. My main thought is that adding status to objects that user interacts with has a higher chance of them seeing the error and taking an action. I do agree that adding it to the workload doesn't feel right.

Comment (mkeeler) [reply internal/catalog/internal/controllers/endpoints/controller.go:314]: This was getting complex enough that I thought it would be best to leave it to a future PR.

**Pull request merged**
